### PR TITLE
txn refactoring: some tidying up

### PR DESCRIPTION
**Pull request by nrc**
###  What have you changed?

The commits stand alone and the descriptions say what they are doing. The first three are fairly trivial tidying up. ~~The forth moves `is_pessimistic_lock` from `Options` to `Command::Prewrite`, which makes sense because that is the only command it is used with.~~

###  What is the type of the changes?

- Engineering (engineering change which doesn't change any feature or fix any issue)

PTAL @MyonKeminta @youjiali1995

### CODE REVIEW

File: src/storage/mod.rs [hunk 1]
```diff
@@ -1645,6 +1647,7 @@ mod tests {
             .async_prewrite(
                 Context::default(),
                 vec![Mutation::Put((Key::from_raw(b"x"), b"100".to_vec()))],
+                vec![false],
```
Comment (sticnarf) [inline src/storage/mod.rs]: Optimistic prewrite needn't provide `is_pessimistic_lock`.

File: src/storage/commands.rs [hunk 2]
```diff
@@ -68,6 +68,8 @@ pub enum CommandKind {
     Prewrite {
         /// The set of mutations to apply.
         mutations: Vec<Mutation>,
+        /// Whether each mutation is locked by a pessimistic lock.
```
Comment (sticnarf) [inline src/storage/commands.rs]: Better to note that this field is only valid for a pessimistic prewrite (for_update_ts > 0)

File: src/storage/kv/mod.rs [hunk 3]
```diff
@@ -56,6 +56,23 @@ pub enum Modify {
     DeleteRange(CfName, Key, Key, bool),
 }
 
+impl Modify {
+    pub fn size(&self) -> usize {
+        let cf = match self {
+            Modify::Delete(cf, _) => cf,
+            Modify::Put(cf, ..) => cf,
+            Modify::DeleteRange(cf, ..) => cf,
```
Comment (sticnarf) [inline src/storage/kv/mod.rs]: this should be also unreachable?

**Comment by nrc** [comment]
@sticnarf addressed comments, PTAL

File: src/server/service/kv.rs [hunk 4]
```diff
@@ -1989,14 +1989,14 @@ fn future_prewrite<E: Engine, L: LockManager>(
...
     options.for_update_ts = req.get_for_update_ts().into();
-    options.is_pessimistic_lock = req.take_is_pessimistic_lock();
     options.txn_size = req.get_txn_size();
     options.min_commit_ts = req.get_min_commit_ts().into();
 
     let (cb, f) = paired_future_callback();
     let res = storage.async_prewrite(
         req.take_context(),
         mutations,
+        req.take_is_pessimistic_lock(),
```
Comment (Little-Wallace) [inline src/server/service/kv.rs]: Could we combine combine is_pessimistic_lock with mutations here? Just like this:
```
    let mutations = req
        .take_mutations()
        .into_iter().zip(is_pessimistic_lock.iter()
        .map(|mut x, is_lock| match x.get_op() {
            Op::Put => Mutation::Put((Key::from_raw(x.get_key()), x.take_value()), is_lock),
            Op::Del => Mutation::Delete(Key::from_raw(x.get_key()), is_lock),
            Op::Lock => Mutation::Lock(Key::from_raw(x.get_key()), is_lock),
            Op::Insert => Mutation::Insert((Key::from_raw(x.get_key()), x.take_value()), is_lock),
            _ => panic!("mismatch Op in prewrite mutations"),
        })
```

Comment (nrc) [reply src/server/service/kv.rs]: I'm doing this in a follow-up PR.

File: src/storage/mod.rs [hunk 5]
```diff
@@ -1700,6 +1703,7 @@ mod tests {
                     Mutation::Put((Key::from_raw(b"b"), b"bb".to_vec())),
                     Mutation::Put((Key::from_raw(b"c"), b"cc".to_vec())),
                 ],
+                vec![false; 3],
```
Comment (sticnarf) [inline src/storage/mod.rs]: @nrc Sorry, there're still a number of places where providing all-false is_pessimistic_locks is not necessary. Would you please remove all of them?

File: src/storage/mvcc/txn.rs [hunk 6]
```diff
@@ -766,7 +751,7 @@ impl<S: Snapshot> MvccTxn<S> {
             ts = commit.prev();
             found_versions += 1;
 
-            if self.write_size >= MAX_TXN_WRITE_SIZE {
+            if self.write_size() >= MAX_TXN_WRITE_SIZE {
```
Comment (youjiali1995) [inline src/storage/mvcc/txn.rs]: This function will be called many times in a gc round.

**Comment by nrc** [comment]
I removed the last commit since I have a better approach that I'm working on. The other comments are all addressed.

**Comment by youjiali1995** [comment]
Please update the PR title and description @nrc

**Pull request merged**